functions/export-csv.js

```diff
@@ -21,13 +21,61 @@ const {db} = require('./common/context');
 
 function exportCsv(req, res) {
   const {
-    project: providedProjectId,
-    featureType: providedFeatureTypeId,
+    project: projectId,
+    featureType: featureTypeId,
     lang: desiredLanguage,
   } = req.query;
 
-  return res.status(500).send('Not supported yet.');
+  let data = {};
+  return db.fetchProject(projectId).then(
+    project => {
+      if (!project.exists) {
+        res.status(404).send('not found');
+        return Promise.reject(new Error('project not found: ' + projectId))
+      }
+      res.type('text/csv');
+      let form = project.get('featureTypes')[featureTypeId]['forms'];
+      let elements = form[Object.keys(form)[0]]['elements'];
+      var arr = []
+      for (var el in elements) {
+        arr.push('"' + escapeQuotes(elements[el]['labels']['_']) + '"');
+      }
+      res.write(arr.join() + '\n');
+      
+      return Promise.all(
+        [db.fetchRecords(project.id, featureTypeId),
+          elements]);
+    }
+  ).then(
+    data => {
+      var records = data[0];
+      var elements = data[1];
+      records.docs.forEach(doc => {
+        var arr = [];
+        if (doc.data()['featureTypeId'] == featureTypeId) {
+          for (var el in elements) {
+            var value = escapeQuotes(doc.data()['responses'][elements[el]['id']]);
+            value = value === undefined ? '' : value;
+            arr.push('"' + value + '"');
+          }
+          res.write(arr.join() + '\n')
+        }
+      })
+
+      return res.end();
+    }
+  ).catch(
+    err => {
+      console.error("=Export Failed= " , err);
+      return res.status(500).end();
+    }
+  )
+
+  function escapeQuotes(str) {
+    return (str + '').replace(/[\\"']/g, '\\$&').replace(/\u0000/g, '\\0');
```

Comment:
The `(str + '')` is where the `undefined` value is coming from. You should provide the default value outside of this function instead; just replace with `str` here.

Reply:
Also, what is the replace(/\u0000/g...) trying to do?

Reply:
Also note that we don't need to escape single quotes either, since strings are delimited. Should only need: `.replace(/"/g, '"')